feat: reject shrinking segment obstacles when generating circle obstacles [PERCEPTION-3288]

plasota opened the pull request:
When driving through an aisle or entering an aisle with low walls, we will first see segment type obstacles. As the wall starts going out of view, these are then (incorrectly) classified as circle type obstacles, which end up shrinking, causing their centers to move toward the robot, causing false hard stops. 

This feature detects when a circle appears close to where a segment previously was in a manner where the endpoints are close on one end and the new segment is colinear with the old segment. If the criteria are met, the conversion to a circle is rejected. As a result, the obstacle is not tracked, and the false positives are mitigated.

plasota committed:
feat: store previous segments to allow comparisons

plasota committed:
feat: reject circles that are shrinking segments

When driving through an aisle or entering an aisle, we will first see segment type obstacles. As the wall starts going out of view, these are then (incorrectly) classified as circle type obstacles, which end up shrinking, causing their centers to move toward the robot, causing false hard stops. This feature detects when a circle appears close to where a segment previously was in a manner where the endpoints are close on one end and the other end is colinear with the old segment. If the criteria are met, the conversion to a cicle is rejected.

plasota committed:
feat: Check if end of shrunk seg falls in old seg

When checking for colinearity of the candidate shrunk segment, make sure the end that moved falls close to between the endpoints of the old segment (up to a given tolerance, as a percentage of the length of the segment)

plasota committed:
feat: Keep buffer of previous segment lists

Addresses potential problem of a segment shrinking -> growing -> shrinking in consecutive frames due to noise and allowing a circle obstacle to be generated

plasota committed:
fix: update comment

dnikitop commented on include/obstacle_detector/utilities/segment.h:
```diff
@@ -107,6 +107,15 @@ class Segment
+    // Checks if point is between segment endpings (within a tolerance to account for noise)
+    Point a = last_point - first_point;
+    Point b = p - first_point;
+    Point c = p - last_point;
```
what is c used for?

plasota replied:
Good catch, I was using it for something else before, and forgot to remove it.

DrCodeWizzard replied:
please don't use single letter variables

plasota replied:
Unused variable removed, single letter variables addressed below.

dnikitop commented on src/obstacle_extractor.cpp:
```diff
@@ -131,6 +136,10 @@ void ObstacleExtractor::pclCallback(const sensor_msgs::PointCloud::ConstPtr pcl_
 
 void ObstacleExtractor::processPoints() {
+  prev_segments_.push_front(move(segments_));
+  if (prev_segments_.size() > p_prev_seg_buffer_size_)
```
normally we don't really use single line if statements, but it seems that this repo has that as part of its style so 👍

DrCodeWizzard replied:
do not use that single line statement. wrap it in { } . you will thank yourself in the future when you need to modify the line or add a debug statement...

plasota replied:
Fixed!

DrCodeWizzard commented on include/obstacle_detector/utilities/segment.h:
```diff
@@ -107,6 +107,15 @@ class Segment
+    Point a = last_point - first_point;
+    Point b = p - first_point;
+    Point c = p - last_point;
+    double t = a.dot(b) / a.lengthSquared();
```
don't use single letter variables. Please make the names more meaningful.

plasota replied:
While I agree in general, other functions in the file use these names for these same vectors. I think keeping the names consistent is better for readability, but I'll also add some comments explaining each variable in this new function.

DrCodeWizzard replied:
Please don't. Don't use the "neighboring" code to affect your "engineering" style. I have worked with a bunch of scientific codes that used "p, q, r" etc as variable names because "it was clear from context what they meant"... Trust me it was not until I opened an algorithm book from which the code was implemented. Do not rely on comments to explain your meaning, as the comments will become stale or get ignored. We have a comment in our codebase that says "DO NOT REMOVE THE LINE BELOW OR SCANNER WILL BREAK" and someone still went and removed the line in question.

DrCodeWizzard commented on include/obstacle_detector/utilities/segment.h:
```diff
@@ -107,6 +107,15 @@ class Segment
+    Point b = p - first_point;
+    Point c = p - last_point;
+    double t = a.dot(b) / a.lengthSquared();
+    return t > 0.0-tol && t < 1.0+tol;
```
inconsistent formatting. use spaces around - and + signs.

plasota replied:
Fixed!

DrCodeWizzard commented on src/obstacle_extractor.cpp:
```diff
@@ -298,6 +307,31 @@ bool ObstacleExtractor::checkSegmentsCollinearity(const Segment& segment, const
 }
 
+bool ObstacleExtractor::checkSegmentShrinking(const Segment& segment, const tf::StampedTransform& transform) {
+  Segment segment_transformed = segment;
```
you are performing a copy over here so you might as well make a copy in the function call itself...

plasota replied:
I've been advised before to always pass by const reference to make it extra clear from the function making the call that the variable will not be changed. Pass by value also implies that, but less explicitly. Also having the variable segment_transformed makes it clear below that the segment has been altered since being passed in.

DrCodeWizzard replied:
Pass by const reference does two things:
1) It works effectively as pass by pointer (so it is sometimes faster than a copy (depending on compiler implementation) as you are just passing a pointer) but allows you to not have to worry about memory management required of pointers.
2) Allows you to bind temporary variables to references (without const, only existing variables not rvalues could be passed).

However in this case you are already doing a copy on the first line of the function, and already paying the cost... so i really don't see the reason for the pass of variable by reference.

DrCodeWizzard commented on src/obstacle_extractor.cpp:
```diff
@@ -298,6 +307,31 @@ bool ObstacleExtractor::checkSegmentsCollinearity(const Segment& segment, const
+    segment_transformed.last_point = transformPoint(segment.last_point, transform);
+  }  
+  vector<const Point*> to_check_for_colinearity;
+  for (auto segment_list : prev_segments_) {
```
auto creates a copy of elements in prev_segments. since each element is a list it is very expensive operation. I recommend doing auto & segment_list which make s a reference to the list. even better you can make it const auto & segment_list...

same comment applies to the line below that goes through segment lists.

plasota replied:
Good catch, fixed!

DrCodeWizzard commented on src/obstacle_extractor.cpp:
```diff
@@ -298,6 +307,31 @@ bool ObstacleExtractor::checkSegmentsCollinearity(const Segment& segment, const
+      // If two endpoints of segment and s are close, check the other side of the segment for colinearity
+      if ((segment_transformed.first_point - s.first_point).length() < p_shrink_end_dist_ ||
+          (segment_transformed.first_point - s.last_point).length() < p_shrink_end_dist_)
+            to_check_for_colinearity.push_back(&segment_transformed.last_point);
```
do not use single line if statements... wrap in {}

plasota replied:
Fixed!

DrCodeWizzard commented on src/obstacle_extractor.cpp:
```diff
@@ -298,6 +307,31 @@ bool ObstacleExtractor::checkSegmentsCollinearity(const Segment& segment, const
+            to_check_for_colinearity.push_back(&segment_transformed.last_point);
+      if ((segment_transformed.last_point - s.first_point).length() < p_shrink_end_dist_ ||
+          (segment_transformed.last_point - s.last_point).length() < p_shrink_end_dist_)
+            to_check_for_colinearity.push_back(&segment_transformed.first_point);
```
same as above

plasota replied:
Fixed!

plasota committed:
fix: requested changes

*Wrap single line if's with brackets
*Fix inconsistent formatting
*Avoid unnecessary copies in loops

plasota committed:
fix: requested change - remove unused variable

plasota committed:
fix: add comments explaining variables

plasota reviewed (commented):
Fixed the style issues for most comments, added some thoughts to others

plasota committed:
fix: comment typo

DrCodeWizzard reviewed (changes requested):
please make the required changes
